service/migration_manager.cc

```diff
@@ -681,35 +681,45 @@ static future<std::vector<mutation>> include_keyspace(
     co_return std::move(mutations);
 }
 
+static future<std::vector<mutation>> do_prepare_new_column_family_announcement(storage_proxy& sp,
+        const keyspace_metadata& ksm, schema_ptr cfm, api::timestamp_type timestamp) {
+    auto& db = sp.local_db();
+    if (db.has_schema(cfm->ks_name(), cfm->cf_name())) {
+        throw exceptions::already_exists_exception(cfm->ks_name(), cfm->cf_name());
+    }
+    if (db.column_family_exists(cfm->id())) {
+        throw exceptions::invalid_request_exception(format("Table with ID {} already exists: {}", cfm->id(), db.find_schema(cfm->id())));
+    }
+
+    mlogger.info("Create new ColumnFamily: {}", cfm);
+
+    return seastar::async([&db, cfm, timestamp] {
+        auto mutations = db::schema_tables::make_create_table_mutations(cfm, timestamp);
+        db.get_notifier().before_create_column_family(*cfm, mutations, timestamp);
+        return mutations;
+    }).then([&sp, &ksm](std::vector<mutation> mutations) {
+        return include_keyspace(sp, ksm, std::move(mutations));
```

Comment:
TBH, I don't know what kind of mutation is included here.

However, we shouldn't call `include_keyspace` in the new `prepare_new_column_family_announcement` anyway. This function is called when the keyspace and the table are created together. We don't have to worry about a node missing a CREATE KEYSPACE migration and we can't assume that the keyspace exists.

I changed the code so that only the old `prepare_new_column_family_announcement` calls `include_keyspace` .

Reply:
> We don't have to worry about a node missing a CREATE KEYSPACE migration and we can't assume that the keyspace exists.

I'm not sure if `include_keyspace` actually assumes that the keyspace exists. I would guess that if the schema definition is not present on disk, it would just add an empty mutation that does nothing.

And I think we **do** have to worry about a potentially missing `CREATE KEYSPACE` -- the issue solved by `include_keyspace` is still potentially present in non-Raft mode, which also uses these `prepare_` functions. (In Raft mode, we indeed don't have to worry.)

Reply:
Thank you for the detailed explanation. I undid the removing of `include_keyspace`.